File: src/codex/middle/PromptCoordinator.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later OR MIT

#include "codex/middle/PromptCoordinator.h"

#include <QUrl>
#include <algorithm>
#include <set>
#include <utility>

namespace codexui::codex::middle {
namespace {

QString text(const std::string &value) {
  return QString::fromUtf8(value.data(), static_cast<qsizetype>(value.size()));
}

std::string stringValue(const nlohmann::json &object, const char *key) {
  if (!object.is_object())
    return {};
  const auto value = object.find(key);
  return value != object.end() && value->is_string() ? value->get<std::string>()
                                                     : std::string{};
}

QString userMessageText(const nlohmann::json &item) {
  QStringList parts;
  const auto content = item.find("content");
  if (content != item.end() && content->is_array()) {
    for (const nlohmann::json &entry : *content) {
      const std::string value = stringValue(entry, "text");
      if (!value.empty())
        parts.push_back(text(value));
    }
  }
  if (parts.empty()) {
    const std::string value = stringValue(item, "text");
    if (!value.empty())
      parts.push_back(text(value));
  }
  return parts.join(QStringLiteral("\n"));
}

QString markdownLinkLabel(QString label) {
  label.replace(QLatin1Char('\\'), QStringLiteral("\\\\"));
  label.replace(QLatin1Char('['), QStringLiteral("\\["));
  label.replace(QLatin1Char(']'), QStringLiteral("\\]"));
  label.replace(QLatin1Char('\r'), QLatin1Char(' '));
  label.replace(QLatin1Char('\n'), QLatin1Char(' '));
  return label;
}

std::vector<std::pair<AuthoritativeItemKey, const ItemPresentation *>>
orderedItems(const std::string &threadId, const ThreadPresentation &thread) {
  std::vector<std::pair<AuthoritativeItemKey, const ItemPresentation *>> result;
  for (const std::string &turnId : thread.turnOrder) {
    const auto turn = thread.turns.find(turnId);
    if (turn == thread.turns.end())
      continue;
    for (const std::string &itemId : turn->second.itemOrder) {
      const auto item = turn->second.items.find(itemId);
      if (item == turn->second.items.end())
        continue;
      result.emplace_back(AuthoritativeItemKey{threadId, turnId, itemId},
                          &item->second);
    }
  }
  return result;
}

} // namespace

QString promptWithFileLinks(QString prompt,
                            std::span<const AttachmentDraft> attachments) {
  QStringList links;
  for (const AttachmentDraft &attachment : attachments) {
    if (attachment.mimeType.startsWith(QStringLiteral("image/")) ||
        attachment.mimeType.startsWith(QStringLiteral("audio/")))
      continue;
    QString target =
        QUrl::fromLocalFile(attachment.path).toString(QUrl::FullyEncoded);
    target.replace(QLatin1Char('['), QStringLiteral("%5B"));
    target.replace(QLatin1Char(']'), QStringLiteral("%5D"));
    target.replace(QLatin1Char('('), QStringLiteral("%28"));
    target.replace(QLatin1Char(')'), QStringLiteral("%29"));
    links.push_back(QStringLiteral("- [%1](%2)")
                        .arg(markdownLinkLabel(attachment.name), target));
  }
  if (links.empty())
    return prompt;
  return prompt + QStringLiteral("\n\nAttached files:\n") +
         links.join(QLatin1Char('\n'));
}

bool PromptSubmission::acceptedTransitionActive(
    qint64 nowMilliseconds) const noexcept {
  return state == PromptState::Accepted && acceptedAtMilliseconds > 0 &&
         nowMilliseconds >= acceptedAtMilliseconds &&
         nowMilliseconds - acceptedAtMilliseconds <
             AcknowledgementTransitionMilliseconds;
}

bool PromptSubmission::localCardVisible(qint64 nowMilliseconds) const noexcept {
  return state == PromptState::Queued || state == PromptState::InFlight ||
         state == PromptState::Failed || !materializedItem ||
         acceptedTransitionActive(nowMilliseconds);
}

std::uint64_t PromptCoordinator::admit(
    std::string threadId, QString prompt,
    std::vector<AttachmentDraft> attachments, nlohmann::json turnOptions,
    const ThreadPresentation *authoritativeThread,
    std::optional<std::string> activeTurnId, qint64 nowMilliseconds) {
  PromptSubmission submission;
  submission.id = nextSubmissionId++;
  submission.admissionOrdinal = nextAdmissionOrdinal++;
  submission.threadId = std::move(threadId);
  submission.clientUserMessageId = "codexui-" +
                                   std::to_string(nowMilliseconds) + '-' +
                                   std::to_string(submission.id);
  submission.prompt = std::move(prompt);
  submission.attachments = std::move(attachments);
  submission.turnOptions = std::move(turnOptions);
  submission.expectedTurnId = std::move(activeTurnId);

  if (authoritativeThread) {
    const auto items = orderedItems(submission.threadId, *authoritativeThread);
    if (!items.empty())
      submission.admissionAnchor = items.back().first;
  }

  const std::uint64_t id = submission.id;
  byThread[submission.threadId].push_back(std::move(submission));
  return id;
}

std::optional<PromptDispatch>
PromptCoordinator::beginNext(const std::string &threadId,
                             std::optional<std::string> activeTurnId) {
  auto found = byThread.find(threadId);
  if (found == byThread.end())
    return std::nullopt;
  if (std::any_of(found->second.begin(), found->second.end(),
                  [](const PromptSubmission &submission) {
                    return submission.state == PromptState::InFlight;
                  }))
    return std::nullopt;
  auto next = std::find_if(found->second.begin(), found->second.end(),
                           [](const PromptSubmission &submission) {
                             return submission.state == PromptState::Queued;
                           });
  if (next == found->second.end())
    return std::nullopt;
  next->state = PromptState::InFlight;
  // Start versus steer is an operation-time fact. A turn which was active
  // when the prompt entered the local queue may have completed meanwhile.
  next->expectedTurnId = std::move(activeTurnId);
  return PromptDispatch{next->id,
                        next->threadId,
                        next->clientUserMessageId,
                        next->prompt,
                        next->attachments,
                        next->turnOptions,
                        next->expectedTurnId};
}

bool PromptCoordinator::acknowledge(
    const std::string &threadId, std::uint64_t submissionId,
    std::optional<std::string> authoritativeTurnId, qint64 nowMilliseconds) {
  PromptSubmission *pending = find(threadId, submissionId);
  if (!pending || pending->state != PromptState::InFlight)
    return false;
  pending->state = PromptState::Accepted;
  pending->acceptedAtMilliseconds = nowMilliseconds;
  pending->error.clear();
  if (authoritativeTurnId)
    pending->expectedTurnId = std::move(authoritativeTurnId);
  return true;
}

bool PromptCoordinator::fail(const std::string &threadId,
                             std::uint64_t submissionId, QString error) {
  PromptSubmission *pending = find(threadId, submissionId);
  if (!pending || (pending->state != PromptState::InFlight &&
                   pending->state != PromptState::Queued))
    return false;
  pending->state = PromptState::Failed;
  pending->error = std::move(error);
  return true;
}

bool PromptCoordinator::requeue(const std::string &threadId,
                                std::uint64_t submissionId) {
  PromptSubmission *pending = find(threadId, submissionId);
  if (!pending || pending->state != PromptState::InFlight)
    return false;
  pending->state = PromptState::Queued;
  return true;
}

std::size_t PromptCoordinator::failQueued(const std::string &threadId,
                                          const QString &error) {
  auto found = byThread.find(threadId);
  if (found == byThread.end())
    return 0;
  std::size_t count = 0;
  for (PromptSubmission &submission : found->second) {
    if (submission.state != PromptState::Queued)
      continue;
    submission.state = PromptState::Failed;
    submission.error = error;
    ++count;
  }
  return count;
}

bool PromptCoordinator::reassignThread(const std::string &fromThreadId,
                                       const std::string &toThreadId) {
  if (fromThreadId == toThreadId)
    return true;
  auto source = byThread.find(fromThreadId);
  if (source == byThread.end())
    return true;
  auto destination = byThread.find(toThreadId);
  const bool sourceInFlight =
      std::any_of(source->second.begin(), source->second.end(),
                  [](const PromptSubmission &submission) {
                    return submission.state == PromptState::InFlight;
                  });
  const bool destinationInFlight =
      destination != byThread.end() &&
      std::any_of(destination->second.begin(), destination->second.end(),
                  [](const PromptSubmission &submission) {
                    return submission.state == PromptState::InFlight;
                  });
  if (sourceInFlight && destinationInFlight)
    return false;

  std::vector<PromptSubmission> moved = std::move(source->second);
  byThread.erase(source);
  for (PromptSubmission &submission : moved) {
    submission.threadId = toThreadId;
    if (submission.admissionAnchor)
      submission.admissionAnchor->threadId = toThreadId;
    if (submission.materializedItem)
      submission.materializedItem->threadId = toThreadId;
  }
  auto &target = byThread[toThreadId];
  target.insert(target.end(), std::make_move_iterator(moved.begin()),
                std::make_move_iterator(moved.end()));
  std::ranges::sort(target, {}, &PromptSubmission::admissionOrdinal);
  return true;
}

void PromptCoordinator::reconcile(
    const std::string &threadId,
    const ThreadPresentation &authoritativeThread) {
  auto found = byThread.find(threadId);
  if (found == byThread.end())
    return;
  const auto items = orderedItems(threadId, authoritativeThread);
  std::set<AuthoritativeItemKey> claimed;
  for (const PromptSubmission &submission : found->second)
    if (submission.materializedItem)
      claimed.insert(*submission.materializedItem);

  for (PromptSubmission &submission : found->second) {
    if (submission.materializedItem)
      continue;

    // A prompt admitted before thread hydration is anchored exactly once when
    // the authoritative tail first becomes available. Later output therefore
    // cannot move the local card through the history.
    if (!submission.admissionAnchor &&
        submission.state == PromptState::Queued && !items.empty())
      submission.admissionAnchor = items.back().first;

    const auto exact = std::ranges::find_if(items, [&](const auto &entry) {
      const auto &[key, item] = entry;
      return !claimed.contains(key) &&
             stringValue(item->raw, "type") == "userMessage" &&
             stringValue(item->raw, "clientId") ==
                 submission.clientUserMessageId;
    });
    if (exact != items.end()) {
      submission.materializedItem = exact->first;
      submission.expectedTurnId = exact->first.turnId;
      claimed.insert(exact->first);
      continue;
    }

    // Semantic acknowledgement remains callback-only. Without protocol
    // client-id support, do not guess from text before that callback arrives.
    if (submission.state != PromptState::Accepted)
      continue;

    std::size_t firstCandidate = 0;
    if (submission.admissionAnchor) {
      const auto anchor = std::ranges::find(
          items, *submission.admissionAnchor,
          [](const auto &entry) -> const AuthoritativeItemKey & {
            return entry.first;
          });
      if (anchor != items.end())
        firstCandidate =
            static_cast<std::size_t>(std::distance(items.begin(), anchor)) + 1;
    }

    for (std::size_t index = firstCandidate; index < items.size(); ++index) {
      const auto &[key, item] = items[index];
      if (submission.expectedTurnId && key.turnId != *submission.expectedTurnId)
        continue;
      if (claimed.contains(key) ||
          stringValue(item->raw, "type") != "userMessage" ||
          userMessageText(item->raw).trimmed() != submission.prompt.trimmed())
        continue;
      submission.materializedItem = key;
      if (!submission.expectedTurnId)
        submission.expectedTurnId = key.turnId;
      claimed.insert(key);
      break;
    }
  }
}

void PromptCoordinator::compactResolved(const std::string &threadId,
                                        qint64 nowMilliseconds) {
  auto found = byThread.find(threadId);
  if (found == byThread.end())
    return;
  for (PromptSubmission &submission : found->second) {
    if (submission.state != PromptState::Accepted ||
        !submission.materializedItem ||
        submission.acceptedTransitionActive(nowMilliseconds) ||
        submission.clientUserMessageId.empty())
      continue;
    std::string{}.swap(submission.clientUserMessageId);
    QString{}.swap(submission.prompt);
    std::vector<AttachmentDraft>{}.swap(submission.attachments);
    submission.turnOptions = nlohmann::json::object();
    submission.error.clear();
    submission.error.squeeze();
    submission.expectedTurnId.reset();
    submission.acceptedAtMilliseconds = 0;
  }
}

std::span<const PromptSubmission>
PromptCoordinator::submissions(const std::string &threadId) const noexcept {
  const auto found = byThread.find(threadId);
  if (found == byThread.end())
    return {};
  return found->second;
}

const PromptSubmission *
PromptCoordinator::submission(const std::string &threadId,
                              std::uint64_t submissionId) const noexcept {
  const auto found = byThread.find(threadId);
  if (found == byThread.end())
    return nullptr;
  const auto candidate =
      std::ranges::find(found->second, submissionId, &PromptSubmission::id);
  return candidate == found->second.end() ? nullptr : &*candidate;
}

bool PromptCoordinator::hasInFlight(
    const std::string &threadId) const noexcept {
  const auto pending = submissions(threadId);
  return std::ranges::any_of(pending, [](const PromptSubmission &submission) {
    return submission.state == PromptState::InFlight;
  });
}

std::vector<std::string> PromptCoordinator::queuedThreadIds() const {
  std::vector<std::string> result;
  for (const auto &[threadId, submissions] : byThread) {
    if (std::ranges::any_of(submissions,
                            [](const PromptSubmission &submission) {
                              return submission.state == PromptState::Queued;
                            }))
      result.push_back(threadId);
  }
  return result;
}

void PromptCoordinator::clearThread(const std::string &threadId) {
  byThread.erase(threadId);
}

PromptSubmission *PromptCoordinator::find(const std::string &threadId,
                                          std::uint64_t submissionId) noexcept {
  auto found = byThread.find(threadId);
  if (found == byThread.end())
    return nullptr;
  auto candidate =
      std::ranges::find(found->second, submissionId, &PromptSubmission::id);
  return candidate == found->second.end() ? nullptr : &*candidate;
}

} // namespace codexui::codex::middle

File: web/src/conversation/PromptCoordinator.ts
import type {ThreadPresentation, ItemPresentation} from "../presentation/PresentationModel.js";
import {isObject, stringMember} from "../presentation/PresentationProtocol.js";
import {AcknowledgementTransitionMilliseconds} from "./MiddleTypes.js";
import type {AuthoritativeItemKey, LocalPromptKey, PromptState} from "./MiddleTypes.js";

export interface AttachmentDraft {path: string; name: string; mimeType: string; size: number}
export interface PromptSubmission {
    id: number; admissionOrdinal: number; threadId: string; clientUserMessageId: string; prompt: string;
    attachments: AttachmentDraft[]; turnOptions: Record<string, unknown>; state: PromptState;
    acceptedAtMilliseconds: number; error: string; admissionAnchor?: AuthoritativeItemKey;
    admissionAtStart: boolean; expectedTurnId?: string; materializedItem?: AuthoritativeItemKey;
}
export interface PromptDispatch {
    id: number; threadId: string; clientUserMessageId: string; prompt: string; attachments: AttachmentDraft[];
    turnOptions: Record<string, unknown>; expectedTurnId?: string;
}
export interface PromptVisualAlias {key: LocalPromptKey; admissionAnchor?: AuthoritativeItemKey; admissionOrdinal: number}
interface RetainedPromptVisualAlias extends PromptVisualAlias {materializedItem: AuthoritativeItemKey}
export interface AuthoritativeItem {key: AuthoritativeItemKey; presentation: ItemPresentation; promptAlias?: PromptVisualAlias}
export interface UserMessageByText {turnId: string; text: string; position: number}
export interface AuthoritativeItemIndex {
    threadId: string; ordered: AuthoritativeItem[]; positions: Map<string, number>;
    userMessagesByClientId: Map<string, number>; userMessagesByText: UserMessageByText[];
}

export function authoritativeKey(key: AuthoritativeItemKey): string {
    return `${key.threadId.length}:${key.threadId}${key.turnId.length}:${key.turnId}${key.itemId.length}:${key.itemId}`;
}
export function authoritativePosition(index: AuthoritativeItemIndex, key: AuthoritativeItemKey): number | undefined {
    return index.positions.get(authoritativeKey(key));
}
function userMessageText(item: unknown): string {
    const parts: string[] = [];
    const content = isObject(item) ? item.content : undefined;
    if (Array.isArray(content)) for (const entry of content) {
        const text = stringMember(entry, "text");
        if (text !== "") parts.push(text);
    }
    if (parts.length === 0) { const fallback = stringMember(item, "text"); if (fallback !== "") parts.push(fallback); }
    return parts.join("\n");
}
export function indexAuthoritativeItems(threadId: string, thread?: ThreadPresentation): AuthoritativeItemIndex {
    const result: AuthoritativeItemIndex = {
        threadId, ordered: [], positions: new Map(), userMessagesByClientId: new Map(), userMessagesByText: [],
    };
    if (!thread) return result;
    for (const turnId of thread.turnOrder) {
        const turn = thread.turns.get(turnId);
        if (!turn) continue;
        for (const itemId of turn.itemOrder) {
            const presentation = turn.items.get(itemId);
            if (!presentation) continue;
            const position = result.ordered.length;
            const key: AuthoritativeItemKey = {kind: "item", threadId, turnId, itemId};
            result.ordered.push({key, presentation});
            result.positions.set(authoritativeKey(key), position);
            if (stringMember(presentation.raw, "type") === "userMessage") {
                const clientId = stringMember(presentation.raw, "clientId");
                if (clientId !== "" && !result.userMessagesByClientId.has(clientId))
                    result.userMessagesByClientId.set(clientId, position);
                const text = userMessageText(presentation.raw).trim();
                result.userMessagesByText.push({turnId: "", text, position}, {turnId, text, position});
            }
        }
    }
    result.userMessagesByText.sort((a, b) => a.turnId.localeCompare(b.turnId) || a.text.localeCompare(b.text) || a.position - b.position);
    return result;
}

export function promptWithFileLinks(prompt: string, attachments: readonly AttachmentDraft[]): string {
    const links = attachments.filter(file => !file.mimeType.startsWith("image/") && !file.mimeType.startsWith("audio/"))
        .map(file => {
            const label = file.name.replaceAll("\\", "\\\\").replaceAll("[", "\\[").replaceAll("]", "\\]")
                .replace(/[\r\n]/gu, " ");
            const target = new URL(`file://${file.path.startsWith("/") ? "" : "/"}${file.path}`).href
                .replaceAll("[", "%5B").replaceAll("]", "%5D").replaceAll("(", "%28").replaceAll(")", "%29");
            return `- [${label}](${target})`;
        });
    return links.length === 0 ? prompt : `${prompt}\n\nAttached files:\n${links.join("\n")}`;
}

export function acceptedTransitionActive(submission: PromptSubmission, now: number): boolean {
    return submission.state === "accepted" && submission.acceptedAtMilliseconds > 0
        && now >= submission.acceptedAtMilliseconds
        && now - submission.acceptedAtMilliseconds < AcknowledgementTransitionMilliseconds;
}
export function localCardVisible(submission: PromptSubmission, now: number): boolean {
    return submission.state === "queued" || submission.state === "inFlight" || submission.state === "failed"
        || submission.materializedItem === undefined || acceptedTransitionActive(submission, now);
}
function cloneKey(key: AuthoritativeItemKey): AuthoritativeItemKey { return {...key}; }

export class PromptCoordinator {
    private readonly byThread = new Map<string, PromptSubmission[]>();
    private readonly visualAliasesByThread = new Map<string, Map<string, RetainedPromptVisualAlias>>();
    private nextSubmissionId = 1;
    private nextAdmissionOrdinal = 1;

    admit(threadId: string, prompt: string, attachments: AttachmentDraft[], turnOptions: Record<string, unknown>,
        authoritativeThread: ThreadPresentation | undefined, activeTurnId: string | undefined, now: number): number {
        const submission: PromptSubmission = {
            id: this.nextSubmissionId++, admissionOrdinal: this.nextAdmissionOrdinal++, threadId,
            clientUserMessageId: `codexui-${now}-${this.nextSubmissionId - 1}`, prompt, attachments: structuredClone(attachments),
            turnOptions: structuredClone(turnOptions), state: "queued", acceptedAtMilliseconds: 0, error: "",
            admissionAtStart: false,
        };
        if (activeTurnId !== undefined) submission.expectedTurnId = activeTurnId;
        if (authoritativeThread) {
            const items = indexAuthoritativeItems(threadId, authoritativeThread);
            const last = items.ordered.at(-1);
            if (last) submission.admissionAnchor = cloneKey(last.key);
        }
        const list = this.byThread.get(threadId) ?? [];
        if (!this.byThread.has(threadId)) this.byThread.set(threadId, list);
        list.push(submission);
        return submission.id;
    }

    beginNext(threadId: string, activeTurnId?: string): PromptDispatch | undefined {
        const list = this.byThread.get(threadId);
        if (!list || list.some(value => value.state === "inFlight")) return undefined;
        const next = list.find(value => value.state === "queued");
        if (!next) return undefined;
        next.admissionAtStart = next.admissionAnchor === undefined;
        next.state = "inFlight";
        if (activeTurnId === undefined) delete next.expectedTurnId; else next.expectedTurnId = activeTurnId;
        const dispatch: PromptDispatch = {
            id: next.id, threadId: next.threadId, clientUserMessageId: next.clientUserMessageId,
            prompt: next.prompt, attachments: structuredClone(next.attachments), turnOptions: structuredClone(next.turnOptions),
        };
        if (next.expectedTurnId !== undefined) dispatch.expectedTurnId = next.expectedTurnId;
        return dispatch;
    }

    acknowledge(threadId: string, id: number, turnId: string | undefined, now: number): boolean {
        const pending = this.find(threadId, id);
        if (!pending || pending.state !== "inFlight") return false;
        pending.state = "accepted"; pending.acceptedAtMilliseconds = now; pending.error = "";
        if (turnId !== undefined) pending.expectedTurnId = turnId;
        return true;
    }
    fail(threadId: string, id: number, error: string): boolean {
        const pending = this.find(threadId, id);
        if (!pending || (pending.state !== "inFlight" && pending.state !== "queued")) return false;
        pending.state = "failed"; pending.error = error; return true;
    }
    requeue(threadId: string, id: number): boolean {
        const pending = this.find(threadId, id);
        if (!pending || pending.state !== "inFlight") return false;
        pending.state = "queued"; pending.admissionAtStart = false; return true;
    }
    failQueued(threadId: string, error: string): number {
        let count = 0;
        for (const pending of this.byThread.get(threadId) ?? []) if (pending.state === "queued") {
            pending.state = "failed"; pending.error = error; ++count;
        }
        return count;
    }
    reassignThread(from: string, to: string): boolean {
        if (from === to) return true;
        const source = this.byThread.get(from);
        const target = this.byThread.get(to) ?? [];
        if (source?.some(value => value.state === "inFlight") && target.some(value => value.state === "inFlight")) return false;
        if (source) {
            this.byThread.delete(from);
            for (const submission of source) {
                submission.threadId = to;
                if (submission.admissionAnchor) submission.admissionAnchor.threadId = to;
                if (submission.materializedItem) submission.materializedItem.threadId = to;
            }
            target.push(...source); target.sort((a, b) => a.admissionOrdinal - b.admissionOrdinal); this.byThread.set(to, target);
        }
        const aliases = this.visualAliasesByThread.get(from);
        if (aliases) {
            this.visualAliasesByThread.delete(from);
            const destination = this.visualAliasesByThread.get(to) ?? new Map();
            for (const alias of aliases.values()) {
                alias.materializedItem.threadId = to;
                if (alias.admissionAnchor) alias.admissionAnchor.threadId = to;
                destination.set(authoritativeKey(alias.materializedItem), alias);
            }
            this.visualAliasesByThread.set(to, destination);
        }
        return true;
    }

    reconcile(threadId: string, threadOrIndex: ThreadPresentation | AuthoritativeItemIndex, now: number): AuthoritativeItemIndex {
        const index = "ordered" in threadOrIndex ? threadOrIndex : indexAuthoritativeItems(threadId, threadOrIndex);
        this.applyVisualAliases(threadId, index);
        const submissions = this.byThread.get(threadId);
        if (!submissions) return index;
        const claimed = index.ordered.map(item => item.promptAlias !== undefined);
        for (const submission of submissions) if (submission.materializedItem) {
            const position = authoritativePosition(index, submission.materializedItem); if (position !== undefined) claimed[position] = true;
        }
        for (const submission of submissions) {
            if (submission.materializedItem) continue;
            if (!submission.admissionAnchor && submission.state === "queued" && index.ordered.length > 0) {
                submission.admissionAnchor = cloneKey(index.ordered.at(-1)!.key); submission.admissionAtStart = false;
            }
            const exact = index.userMessagesByClientId.get(submission.clientUserMessageId);
            if (exact === undefined || claimed[exact]) continue;
            submission.materializedItem = cloneKey(index.ordered[exact]!.key);
            submission.expectedTurnId = submission.materializedItem.turnId;
            claimed[exact] = true;
        }
        for (const submission of submissions) {
            if (submission.materializedItem || submission.state !== "accepted") continue;
            let first = 0;
            if (submission.admissionAnchor) first = (authoritativePosition(index, submission.admissionAnchor) ?? -1) + 1;
            const turnId = submission.expectedTurnId ?? "";
            const candidate = index.userMessagesByText.find(value => value.turnId === turnId
                && value.text === submission.prompt.trim() && value.position >= first && !claimed[value.position]);
            if (!candidate) continue;
            submission.materializedItem = cloneKey(index.ordered[candidate.position]!.key);
            submission.expectedTurnId ??= submission.materializedItem.turnId;
            claimed[candidate.position] = true;
        }
        const aliases = this.visualAliasesByThread.get(threadId) ?? new Map<string, RetainedPromptVisualAlias>();
        for (const submission of submissions) {
            if (submission.state !== "accepted" || !submission.materializedItem || acceptedTransitionActive(submission, now)) continue;
            aliases.set(authoritativeKey(submission.materializedItem), {
                key: {kind: "prompt", submissionId: submission.id}, admissionOrdinal: submission.admissionOrdinal,
                materializedItem: cloneKey(submission.materializedItem),
                ...(submission.admissionAnchor ? {admissionAnchor: cloneKey(submission.admissionAnchor)} : {}),
            });
        }
        this.visualAliasesByThread.set(threadId, aliases);
        this.byThread.set(threadId, submissions.filter(submission => submission.state !== "accepted"
            || !submission.materializedItem || acceptedTransitionActive(submission, now)));
        this.applyVisualAliases(threadId, index);
        return index;
    }

    submissions(threadId: string): readonly PromptSubmission[] { return this.byThread.get(threadId) ?? []; }
    submission(threadId: string, id: number): PromptSubmission | undefined { return this.find(threadId, id); }
    hasInFlight(threadId: string): boolean { return this.submissions(threadId).some(value => value.state === "inFlight"); }
    queuedThreadIds(): string[] {
        return [...this.byThread].filter(([, values]) => values.some(value => value.state === "queued")).map(([id]) => id).sort();
    }
    clearThread(threadId: string): void { this.byThread.delete(threadId); this.visualAliasesByThread.delete(threadId); }
    private find(threadId: string, id: number): PromptSubmission | undefined {
        return this.byThread.get(threadId)?.find(value => value.id === id);
    }
    private applyVisualAliases(threadId: string, index: AuthoritativeItemIndex): void {
        const aliases = this.visualAliasesByThread.get(threadId);
        if (!aliases) return;
        for (const [key, retained] of aliases) {
            const position = index.positions.get(key);
            if (position !== undefined) index.ordered[position]!.promptAlias = retained;
        }
    }
}
